### DISCUSSION TITLE
Unary transit time (labour times) for VRP

### DISCUSSION THREAD

**Discussion by stevenshack** [google/or-tools#2720] [Routing (and legacy CP) questions]
**What language and solver does this apply to?**
This applies to VRP problems in all languages.

**Describe the problem you are trying to solve.**

Currently or-tools VTP doesn't support having both a transit and a unary callback.
As suggested in threads like this: https://github.com/google/or-tools/issues/498
the recommended approach is to add on service time to transit time directly in ones
time matrix. This makes for error prone book-keeping and time conversions when interfacing
with other parts of code a planner may interact with. 

**Describe the solution you'd like**

I would like the support of unary service times in addition to transit times. 

**Describe alternatives you've considered**
Hand book-keeping. adding and subtracting service times within the various pieces of code. 

**Additional context**
None.

**Comment by Mizux** [comment]
If you looking for a overload of AddDimension() or SetArcCost() we won't do it, sorry.
It will only increase our technical debt to maintain without any adding value.

Otherwise you can already:
* create a transit only or service time only dimension (in addition or not of a dimension with both value)
* Instead of creating a matrix you can perform the sum directly in the callback keeping the data separate if it easier for you
  e.g.
  ```python
  # Data model with service time and transit time matrix separated.
  data['service_time'] = [...]
  data['transit_time'] = [[...]]

  def time_evaluator(from_index, to_index):
    """Returns the total time between the two nodes"""
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return data['service_time'][from_node] +  data['transit_time'][from_node][to_node]
  ```
* BTW, when dealing with TW you'll need both durations on the same duration to accurately compute time...
* don't hesitate to wrap the or-tools stuff in your own abstraction class to hide this complexity from the rest of the code and improve your QoL
  e.g. create a library, "routes optimization", with your ideal API which internally call or-tool's routing

**Comment by stevenshack** [comment]
Okay. I understand your point here. Let me try and tackle this a different way.

How difficult would it be to add  pickup delivery sets (rather than just pairs), that are to be transited sequentially? 

Say I have a pickup A and delivery B. with some service time cost associated with both of them.
 I could setup a dummy node C with a transit cost from A to B equal to my service time  cost, and infinity everywhere else. Then my "Pickup/Delivery is A(at time window, no mudging)-> C (imposing transit cost)-> B(at time window, no mudging).

Can Pickup delivery sets/routes be added sanely?

**Comment by Mizux** [comment]
ok, so question is "How to add a Pickup/Delivery pair when the arc Pickup->Delivery is forbidden ?"

Honestly, never tested...
